### ISSUE TITLE
Write an istanbul verifyConditions plugin

### ISSUE THREAD

**Issue by boennemann** [semantic-release/semantic-release#68]
The popular code coverage generator [istanbul](https://github.com/gotwarlost/istanbul) offers the [`check-coverage`](https://github.com/gotwarlost/istanbul#the-check-coverage-command) command.

I'd love to see a semantic-release verifyConditions plugin (just like [condition-travis](https://github.com/semantic-release/condition-travis)) that aborts any release where defined coverage thresholds aren't met, but I might not immediately have the time to do it myself.

If you want to give this a shot let me know in this issue and I'm happy to help wherever I can. You can reach me in the [semantic-release gitter room](https://gitter.im/semantic-release/semantic-release), or on [Twitter](https://twitter.com/boennemann).

**Comment by kentcdodds** [comment]
For what it's worth, I actually use this as part of my scripts and pre-commit hooks so the build will break if a PR or commit doesn't have sufficient code coverage.

**Comment by ariporad** [comment]
Me too. I think most code coverage tools have an option to fail if they
don't pass a certain threshold.


Ari

**Comment by apowers313** [comment]
Seeing as it's automated, I don't see any harm in doing it in semantic-release too.

Istanbul has a `json-summary` reporter that kicks out a report in the following format:
``` json
{
  "total": {
    "lines": {
      "total": 3,
      "covered": 3,
      "skipped": 0,
      "pct": 100
    },
    "statements": {
      "total": 3,
      "covered": 3,
      "skipped": 0,
      "pct": 100
    },
    "functions": {
      "total": 1,
      "covered": 1,
      "skipped": 0,
      "pct": 100
    },
    "branches": {
      "total": 0,
      "covered": 0,
      "skipped": 0,
      "pct": 100
    },
    "linesCovered": {
      "7": 3,
      "8": 3,
      "9": 3
    }
  },
  "/private/tmp/open-element-template/test.js": {
    "lines": {
      "total": 3,
      "covered": 3,
      "skipped": 0,
      "pct": 100
    },
    "statements": {
      "total": 3,
      "covered": 3,
      "skipped": 0,
      "pct": 100
    },
    "functions": {
      "total": 1,
      "covered": 1,
      "skipped": 0,
      "pct": 100
    },
    "branches": {
      "total": 0,
      "covered": 0,
      "skipped": 0,
      "pct": 100
    },
    "linesCovered": {
      "7": 3,
      "8": 3,
      "9": 3
    }
  }
}
```

Making this work would require:
1. Using Istanbul for code coverage
2. Ensuring that `json-reporter` was enabled as a reporter
3. Having a configuration option for the plugin that points to the `coverage` directory
4. Having a configuration option for the passing threshold 

The semantic-report plugin would read in `{{coverage}}/coverage-summary.json`, compare `{{threshold}}` to the total coverage for lines, statements, functions and branches and return `true` if the coverage is greater than or equal to the threshold.

One question -- is there a way for a plugin to send a failure notification (email, [slack](https://slack.com/), etc.) so that a minimal amount of time is spent trying to understand why a release didn't go out? Doesn't seem like every plugin should write their own notification mechanisms...

**Comment by apowers313** [comment]
Is the assumption for `verifyConditions` that if the callback has an error the condition fails; otherwise, the condition succeeds?

**Comment by boennemann** [comment]
Here are several open things and I'll try to address these now :)

1. I think it makes sense to run this only at release time and not during pre-commit, because this is exactly where it is relevant. Enforcing passing tests/sufficient coverage at commit time makes it impossible to push WIP things and let them run on CI, you can't push failing tests to prove they're failing, this is unfriendly for newcomers, who might get hindered from sending a PR in the first place etc.

2. Istanbul already has this built in:
```Available commands are:

      check-coverage
              checks overall/per-file coverage against thresholds from coverage
              JSON files. Exits 1 if thresholds are not met, 0 otherwise
```

https://github.com/gotwarlost/istanbul#the-command-line

3. At this point I'm wondering if it makes sense to only implement a  `run-script` plugin (#69) instead, because of how straight forward `istanbul check-coverage` already is

4. If the  `verifyConditions` callback gets an error passed as the first argument it is considered a failure

5. I'm thinking of building a mechanism where the results of the verification get attached to the commit via the GitHub status API.


Best,
Stephan

**Comment by apowers313** [comment]
If you were to implement Istanbul as a verifyConditions script, wouldn't that mean that you would have to run the test suite twice -- once during Travis's `npm test` and once during semantic-release's `verifyConditiions`? If that's the case, it might be too much overhead for large test suites...

I like checking Istanbul coverage during semantic-release because 1) it won't mark the build as failing if coverage thresholds aren't met; 2) all of the release requirements are managed by a single decisioning engine, rather than trying to figure what is breaking and where; 3) it simply stops a release if coverage is insufficient, which seems like the most appropriate (and semantically correct) action.

**Comment by kentcdodds** [comment]
If I need a WIP commit, I simply commit with `--no-verify` and it skips the commit hooks. I actually prefer the build to break if the thresholds aren't met.

Sounds to me like this would be a fine micro-plugin. I personally wouldn't use it.

**Comment by satya164** [comment]
[nyc](https://github.com/bcoe/nyc) will be a better choice as it also supports test runner which run tests in subprocesses.

**Comment by boennemann** [comment]
@satya164 nyc uses istanbul under the hood, so this will work with coverage reports nyc has reported, too.

**Comment by saiichihashimoto** [comment]
I'm really confused as to the value of this (or #67, for that matter). From what I gather, semantic-release is about automating package publishing. And in continuous integration, it makes sense for the build to fail before we reach this point if our tests fail. Don't we want the tests (and ultimately, the build) to fail if the coverage is too low?

Perhaps the purpose of semantic-release isn't defined well enough. From what I understand, semantic release:

- Examines the package, determining if it's the right situation to release and, if it is, determining the version to release.
- Doing the actual release
- Creating documentation and git tags

It doesn't look like semantic-release should reinvent running tests. That can (and currently should be) run before semantic-release and fail the build if they don't pass. The exceptions to this would be running cracks or #65 but even those are running tests to determine what the next version should be: a job of semantic-release.

**TLDR** If coverage should abort the release, isn't the best way to fail the build, just as failing tests would?

**Comment by phra** [comment]
@saiichihashimoto i'm exactly doing that. in an agile environment, every commit should be able to be promoted to master without disasters, so if the coverage is not correct, the commit itself should fail.

**Comment by pvdlg** [comment]
As mentioned in several comments running test coverage is more appropriate in the test phase than in the release phase. If we choose to enforce a certain percentage of coverage, and it's not met the test should fails and `semantic-release` release shouldn't even be called.

As anyone any objection regarding closing this issue?

**Comment by saiichihashimoto** [comment]
Let's close it
